vm/vmm/vmm.go

```diff
@@ -0,0 +1,422 @@
+
+	var tee io.Writer
+	if inst.debug {
+		tee = os.Stdout
+	}
+	inst.merger = vmimpl.NewOutputMerger(tee)
+
+	inst.console()
+
+	var bootOutput []byte
+	bootOutputStop := make(chan bool)
+	go func() {
+		for {
+			select {
+			case out := <-inst.merger.Output:
+				bootOutput = append(bootOutput, out...)
+			case <-bootOutputStop:
+				bootOutputStop <- true
+				return
+			}
+		}
+	}()
+
+	done := func() {
+		bootOutputStop <- true
+		<-bootOutputStop
+		close(bootOutputStop)
+	}
+
+	// Wait for ssh server to come up.
+	time.Sleep(5 * time.Second)
+	start := time.Now()
+	for {
+		c, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%v", inst.sshhost, inst.sshport), 1*time.Second)
+		if err == nil {
+			c.SetDeadline(time.Now().Add(1 * time.Second))
+			var tmp [1]byte
+			n, err := c.Read(tmp[:])
+			c.Close()
+			if err == nil && n > 0 {
+				break // ssh is up and responding
+			}
+			time.Sleep(3 * time.Second)
+		}
+		if time.Since(start) > 2*time.Minute {
+			done()
+			inst.merger.Wait()
+			return vmimpl.BootError{Title: "ssh server did not start", Output: bootOutput}
+		}
+	}
+	done()
+
+	return nil
+}
+
+func (inst *instance) Close() {
+	if out, err := inst.vmctl("stop", inst.vmIdent(), "-f"); err != nil {
+		if inst.debug {
+			log.Logf(0, "vmctl stop %s: %v: %s", inst.vmIdent(), err, string(out))
+		}
+	}
+
+	os.Remove(inst.image)
+}
+
+func (inst *instance) Forward(port int) (string, error) {
+	addr := fmt.Sprintf("100.64.%v.2:%v", inst.vmID, port)
+	return addr, nil
+}
+
+func (inst *instance) Copy(hostSrc string) (string, error) {
+	vmDst := filepath.Join("/root", filepath.Base(hostSrc))
+	args := append(vmimpl.SCPArgs(inst.debug, inst.sshkey, inst.sshport), hostSrc, inst.sshuser+"@"+inst.sshhost+":"+vmDst)
+	cmd := osutil.Command("scp", args...)
+	if inst.debug {
+		log.Logf(0, "running command: scp %#v", args)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stdout
+	}
+	if err := cmd.Start(); err != nil {
+		return "", err
+	}
+	done := make(chan bool)
+	go func() {
+		select {
+		case <-time.After(3 * time.Minute):
+			cmd.Process.Kill()
+		case <-done:
+		}
+	}()
+	err := cmd.Wait()
+	close(done)
+	if err != nil {
+		return "", err
+	}
+	return vmDst, nil
+}
+
+func (inst *instance) Run(timeout time.Duration, stop <-chan bool, command string) (
+	<-chan []byte, <-chan error, error) {
+	rpipe, wpipe, err := osutil.LongPipe()
+	if err != nil {
+		return nil, nil, err
+	}
+	inst.merger.Add("ssh", rpipe)
+
+	args := append(vmimpl.SSHArgs(inst.debug, inst.sshkey, inst.sshport), inst.sshuser+"@"+inst.sshhost, command)
+	if inst.debug {
+		log.Logf(0, "running command: ssh %#v", args)
+	}
+	cmd := osutil.Command("ssh", args...)
+	cmd.Stdout = wpipe
+	cmd.Stderr = wpipe
+	if err := cmd.Start(); err != nil {
+		wpipe.Close()
+		return nil, nil, err
+	}
+	wpipe.Close()
+	errc := make(chan error, 1)
+	signal := func(err error) {
+		select {
+		case errc <- err:
+		default:
+		}
+	}
+
+	go func() {
```

Comment:
Please check if `vmimpl.Multiplex` does what you need here.
Somebody needs to call inst.merger.Wait on all paths.

Reply:
Since the console pipes can change I didn't manage to make use of `Multiplex` just yet. However, I did add the missing calls to `inst.merger.Wait()`.